cpp/src/filling/sequence.cu

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/filling.hpp>
+#include <cudf/types.hpp>
+#include <cudf/utilities/error.hpp>
+#include <cudf/scalar/scalar.hpp>
+#include <cudf/scalar/scalar_device_view.cuh>
+
+#include <cudf/column/column_factories.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/detail/iterator.cuh>
+
+namespace cudf {
+namespace experimental {
+
+namespace detail {
+
+namespace {
+
+// This functor only exists here because using a lambda directly in the tabulate() call generates the cryptic
+// __T289 link error.  This seems to be related to lambda usage within functions using SFINAE.
+template<typename T>
+struct tabulator {
+   cudf::numeric_scalar_device_view<T> const n_init;
+   cudf::numeric_scalar_device_view<T> const n_step;
+
+   T __device__ operator()(cudf::size_type i){
+      return n_init.value() + (i * n_step.value());
+   }
+};
+
+/**
+ * @brief Functor called by the `type_dispatcher` to generate the sequence specified 
+ * by init and step.
+ */
+struct sequence_functor {   
+   template <typename T, typename std::enable_if_t<cudf::is_numeric<T>() and not cudf::is_boolean<T>()>* = nullptr>
+   std::unique_ptr<column> operator()(size_type size, scalar const& init, scalar const& step,
+                                      rmm::mr::device_memory_resource* mr,
+                                      cudaStream_t stream)
+   {
+      auto result = make_fixed_width_column(init.type(), size, mask_state::UNALLOCATED, stream, mr);
+      auto result_device_view = mutable_column_device_view::create(*result, stream);      
+                  
+      auto n_init = get_scalar_device_view( static_cast<cudf::experimental::scalar_type_t<T>&>(const_cast<scalar&>(init)) );
```

Comment:
Why are we `const_cast`ing? That's rarely a good idea.

Reply:
Unfortunately, get_scalar_device_view() requires a non-const reference.  Since there's no non-mutable scalar device view, even though we're only ever reading it, we have to treat it as if it were non-const.   I could change things so that we're not passing const& but then we're just propagating the issue.  I figured keeping it isolated in this one spot was preferable.

There's another example of this here:

https://github.com/rapidsai/cudf/blob/eac284d91163de4f8feda1d27bb808d760bb1509/cpp/src/strings/substring.cu#L135-L137